src/data/representation.lean

```diff
@@ -0,0 +1,312 @@
+structure, while in points 2. and 4. the presented data determines the k-module structure.
+For that matter, at this point we haven't even described the equivalence
+`module k M ≃ (k →* (M →+ M))`.
+In this file we define the typeclass `representation k G M`, and provide the functions
+```
+representation.as_monoid_hom [representation k G M] : G →* (M →ₗ[k] M)
+of_monoid_hom (ρ : G →* (M →ₗ[k] M)) : representation k G M
+```
+(addressing the equivalence between 1. and 3. above)
+and the functions
+```
+extend_scalars [representation k G M] : module (monoid_algebra k G) M
+representation.of_monoid_algebra_module [module (monoid_algebra k G) M] : representation k G M
+```
+(where the second function provides the `representation k G M` relative to the `module k M`
+coming from `restrict_scalars k (monoid_algebra k G) M`).
+These provide the equivalence between 1. and 2. above.
+
+The function `monoid_algebra.lift` in `data.monoid_algebra` then provides the equivalence
+between 3. and 5.
+
+(There's obviously a quite direct connection between points 2. and 5. above,
+which hasn't been developed yet.)
+
+(We don't yet say anything about point 4. above; perhaps it's less useful.)
+
+Finally, note that really all of these equivalences should be done as equivalences of categories,
+not just functions back and forth.
+We'll get there in a later file!
+
+## Tags
+
+linear action, representation
+-/
+
+universes u v w
+
+section prio
+set_option default_priority 100 -- see Note [default priority]
+/--
+A k-linear representation of a group G.
+-/
+@[nolint dangerous_instance] -- k is an `out_param`, so this is a false positive.
+class representation
+  (k : out_param $ Type v) [comm_ring k] (G : Type u) [monoid G] (M : Type w) [add_comm_group M] [module k M]
+  extends distrib_mul_action G M : Type (max u v w) :=
+(smul_smul : ∀ (g : G) (r : k) (m : M), g • (r • m) = r • (g • m))
+end prio
+
+section
+variables {k : Type v} {G : Type u} {M : Type w}
+variables [comm_ring k] [monoid G] [add_comm_group M] [module k M]
+
+/-- A group element acts by an k-linear map in any k-linear representation. -/
+def smul.linear_map [representation k G M] (g : G) : M →ₗ[k] M :=
+{ to_fun := λ m, g • m,
+  add := λ m m', by simp [smul_add],
+  smul := λ r m, by simp [representation.smul_smul], }
+
+@[simp]
+lemma smul.linear_map_apply [representation k G M] (g : G) (m : M) :
+  (smul.linear_map g : M →ₗ[k] M) m = g • m := rfl
+end
+
+namespace representation
+variables (k : Type v) (G : Type u)
+variables [comm_ring k] [monoid G]
+variables (M : Type w) [add_comm_group M] [module k M]
+
+/--
+A k-linear representation of G on M can be thought of as
+a multiplicative map from G into the k-linear endomorphisms of M.
+-/
+def as_monoid_hom [representation k G M] : G →* (M →ₗ[k] M) :=
+{ to_fun := λ g, smul.linear_map g,
+  map_one' := by { ext, simp, },
+  map_mul' := λ g g', by { ext, simp [mul_smul], }, }
```

Comment:
In the Langlands programme it is very common to let pi be a representation of G on V. Note that pi is not G or V, it's the name of the structure in question, although it has a coercion to a function G -> End(V). Currently this function is called `as_monoid_hom [representation k G M]` in Lean. Giving this data is equivalent to giving the representation and pi is a very common interface in modern representation theory.